pkg/discovery/worker/k8s_discovery_worker.go

```diff
@@ -0,0 +1,243 @@
+		for _, smWorker := range resourceMonitoringWorkers {
+			// Assign task to monitoring worker.
+			smWorker.ReceiveTask(clusterMonitorTask)
+			monitoringSink := smWorker.Do()
+			// Don't do any filtering
+			worker.sink.MergeSink(monitoringSink, nil)
+		}
+	}
+
+	var discoveryResult []*proto.EntityDTO
+	// Build EntityDTO
+	// node
+	stitchingManager := stitching.NewStitchingManager(worker.config.stitchingPropertyType)
+
+	var nodes []runtime.Object
+	nodeNameUIDMap := make(map[string]string)
+	for _, n := range currTask.NodeList() {
+		node := n
+		nodes = append(nodes, node)
+		nodeNameUIDMap[node.Name] = string(node.UID)
+		stitchingManager.StoreStitchingValue(node)
+	}
+
+	nodeEntityDTOBuilder := dtofactory.NewNodeEntityDTOBuilder(worker.sink, stitchingManager)
+	nodeEntityDTOs, err := nodeEntityDTOBuilder.BuildEntityDTOs(nodes)
+	if err != nil {
+		glog.Errorf("Error while creating node entityDTOs: %v", err)
+		// TODO Node discovery fails, directly return?
+	}
+	glog.V(2).Infof("Build %d node entityDTOs.", len(nodeEntityDTOs))
+	discoveryResult = append(discoveryResult, nodeEntityDTOs...)
+
+	// pod
+	pods := worker.findPodsRunningInTaskNodes(currTask.NodeList())
+	podEntityDTOBuilder := dtofactory.NewPodEntityDTOBuilder(worker.sink, stitchingManager, nodeNameUIDMap)
+	podEntityDTOs, err := podEntityDTOBuilder.BuildEntityDTOs(pods)
+	if err != nil {
+		glog.Errorf("Error while creating pod entityDTOs: %v", err)
+		// TODO Pod discovery fails, directly return?
+	}
+	discoveryResult = append(discoveryResult, podEntityDTOs...)
+	glog.V(2).Infof("Build %d pod entityDTOs.", len(podEntityDTOs))
+
+	// application
+	applicationEntityDTOBuilder := dtofactory.NewApplicationEntityDTOBuilder(worker.sink)
+	appEntityDTOs, err := applicationEntityDTOBuilder.BuildEntityDTOs(pods)
+	if err != nil {
+		glog.Errorf("Error while creating application entityDTOs: %v", err)
+		// TODO Application discovery fails, return?
+	}
+	discoveryResult = append(discoveryResult, appEntityDTOs...)
+	glog.V(2).Infof("Build %d application entityDTOs.", len(appEntityDTOs))
+
+	// Send result
+	glog.V(3).Infof("Discovery result of worker %s is: %++v", worker.id, discoveryResult)
+
+	result := task.NewTaskResult(worker.id, task.TaskSucceeded).WithContent(discoveryResult)
+	return result
+}
+
+// Discover pods running on nodes specified in task.
+func (worker *k8sDiscoveryWorker) findPodsRunningInTaskNodes(nodes []*api.Node) []runtime.Object {
+	var podList []runtime.Object
+	for _, node := range nodes {
+		nodeNonTerminatedPodsList, err := worker.findRunningPods(node.Name)
+		if err != nil {
+			glog.Errorf("Failed to find non-ternimated pods in %s", node.Name)
+			continue
+		}
+		for _, p := range nodeNonTerminatedPodsList.Items {
+			pod := p
+			podList = append(podList, &pod)
+		}
+	}
+	return podList
+}
+
+func (worker *k8sDiscoveryWorker) findRunningPods(nodeName string) (*api.PodList, error) {
```

Comment:
running pods list API is called four times:
   (1) pkg/discovery/monitoring/k8sconntrack/k8sconntrack_monitor.go
   (2) pkg/discovery/monitoring/master/cluster_info_scraper.go           
   (3)  pkg/discovery/worker/k8s_discovery_worker.go
   (4) pkg/discovery/worker/service_discovery_worker.go  
            ---buildPodClusterIDToPod()

  1. Shall we cache the Pods?
     cache the Pods by Node, or use a cache with index by node Name.
Here is a cache with index:
https://github.com/kubernetes/client-go/blob/master/tools/cache/thread_safe_store.go

  2. Seems you consider a lot about flexibility that new types of metrics scrapers can be added easily?
      However, from my understanding, there will not be too many kind of scrapers, nor will then change much. And even you can add new types of scrapers easily, you have to parse the new metrics, and change the code about buildDTOs.
     So my suggestion is to store the different metrics separately.  Let another stage of the Pipeline use all these kinds of metrics to build DTOs.